File: services/order/internal/repository/order.go
package repository

import (
	"context"
	"errors"
	"time"
)

// Sentinel Errors
var (
	ErrOrderNotFound           = errors.New("order not found")
	ErrOrderNotCancellable     = errors.New("order cannot be cancelled in its current status")
	ErrDuplicateIdempotencyKey = errors.New("idempotency key already exists")
	ErrInvalidPaginationCursor = errors.New("invalid pagination cursor")
)

type OrderSide string
type OrderType string
type OrderStatus string

const (
	SideBuy  OrderSide = "BUY"
	SideSell OrderSide = "SELL"

	TypeLimit  OrderType = "LIMIT"
	TypeMarket OrderType = "MARKET"

	StatusOpen            OrderStatus = "OPEN"
	StatusPartiallyFilled OrderStatus = "PARTIALLY_FILLED"
	StatusFilled          OrderStatus = "FILLED"
	StatusCancelling      OrderStatus = "CANCELLING"
	StatusCancelled       OrderStatus = "CANCELLED"
)

// Order represents a user's trading order record.
type Order struct {
	ID                string
	UserID            string
	MarketID          string
	Side              OrderSide
	OrderType         OrderType
	Price             *string // nil for MARKET orders without price cap
	Quantity          string
	FilledQuantity    string
	RemainingQuantity string
	Status            OrderStatus
	IdempotencyKey    *string
	CreatedAt         time.Time
	UpdatedAt         time.Time
}

// OrderRepository defines the database operations for Order Service.
type OrderRepository interface {
	OutboxRepository

	// FindByIdempotencyKey looks up an order by client-supplied idempotency key. Returns (nil, nil) if not found.
	FindByIdempotencyKey(ctx context.Context, key string) (*Order, error)

	// CreateOrder inserts the order and outbox record atomically in one transaction.
	CreateOrder(ctx context.Context, o *Order, outboxPayload []byte) error

	// GetByID retrieves a single order by order ID and user ID.
	GetByID(ctx context.Context, orderID, userID string) (*Order, error)

	// UpdateStatusToCancelling marks order status as CANCELLING and adds Outbox event.
	UpdateStatusToCancelling(ctx context.Context, o *Order, outboxPayload []byte) error

	// ListOrders returns a paginated list of orders matching filters.
	ListOrders(ctx context.Context, userID, marketID, cursor string, side OrderSide, status OrderStatus, fromTime, toTime *time.Time, limit int32) ([]*Order, error)
}

File: frontend/src/api/order.ts
import client from './client'

export interface CreateOrderRequest {
  market_id: string
  side: 'BUY' | 'SELL'
  order_type: 'LIMIT' | 'MARKET'
  price?: string
  quantity: string
}

export interface Order {
  id: string
  user_id: string
  market_id: string
  side: string
  order_type: string
  status: string
  price: string
  quantity: string
  filled_quantity: string
  created_at: string
  updated_at: string
}

export interface ListOrdersParams {
  market_id?: string
  status?: string
  limit?: number
}

export const orderApi = {
  // POST /api/v1/orders — Create a new limit/market order
  createOrder: async (data: CreateOrderRequest): Promise<Order> => {
    const res = await client.post<Order>('/api/v1/orders', data)
    return res.data
  },

  // GET /api/v1/orders — List user orders
  listOrders: async (params?: ListOrdersParams): Promise<Order[]> => {
    const res = await client.get<{ orders: Order[] }>('/api/v1/orders', { params })
    return res.data?.orders || []
  },

  // GET /api/v1/orders/{id} — Get single order
  getOrder: async (id: string): Promise<Order> => {
    const res = await client.get<Order>(`/api/v1/orders/${id}`)
    return res.data
  },

  // POST /api/v1/orders/{id}/cancel — Cancel an open order
  cancelOrder: async (id: string): Promise<Order> => {
    const res = await client.post<Order>(`/api/v1/orders/${id}/cancel`)
    return res.data
  },
}
